Build failure: magma

### Steps To Reproduce

Steps to reproduce the behavior:
1. build magma

### Build log

From [the full log](https://github.com/NixOS/nixpkgs/files/10933692/nix-log.txt):

```
[2813/3430] Linking CXX shared library lib/libmagma.so
FAILED: lib/libmagma.so
: && /nix/store/ds6ivg31k3l0pjhhf3s769bkpmafa54g-gcc-wrapper-11.3.0/bin/c++ -fPIC -std=c++11 -fopenmp -Wall -Wno-unused-function -O3 -DNDEBUG   -shared -Wl,-soname,libmagma
/nix/store/76l4v99sk83ylfwkz8wmwrm4s8h73rhd-glibc-2.35-224/lib/crti.o: in function `_init':
(.init+0xb): relocation truncated to fit: R_X86_64_REX_GOTPCRELX against undefined symbol `__gmon_start__'
CMakeFiles/magma.dir/control/get_batched_crossover.cpp.o: in function `magma_use_zgeqrf_batched_fused_update':
get_batched_crossover.cpp:(.text+0x23a): relocation truncated to fit: R_X86_64_REX_GOTPCRELX against symbol `zgeqrf_panel_decision_a100' defined in .bss section in CMakeFil
CMakeFiles/magma.dir/control/get_batched_crossover.cpp.o: in function `magma_use_cgeqrf_batched_fused_update':
get_batched_crossover.cpp:(.text+0x31a): relocation truncated to fit: R_X86_64_REX_GOTPCRELX against symbol `cgeqrf_panel_decision_a100' defined in .bss section in CMakeFil
CMakeFiles/magma.dir/control/get_batched_crossover.cpp.o: in function `magma_use_dgeqrf_batched_fused_update':
get_batched_crossover.cpp:(.text+0x3fa): relocation truncated to fit: R_X86_64_REX_GOTPCRELX against symbol `dgeqrf_panel_decision_a100' defined in .bss section in CMakeFil
CMakeFiles/magma.dir/control/get_batched_crossover.cpp.o: in function `magma_use_sgeqrf_batched_fused_update':
get_batched_crossover.cpp:(.text+0x4da): relocation truncated to fit: R_X86_64_REX_GOTPCRELX against symbol `sgeqrf_panel_decision_a100' defined in .bss section in CMakeFil
CMakeFiles/magma.dir/control/get_batched_crossover.cpp.o: in function `__static_initialization_and_destruction_0(int, int) [clone .constprop.0]':
get_batched_crossover.cpp:(.text.startup+0xce9): relocation truncated to fit: R_X86_64_REX_GOTPCRELX against symbol `sgeqrf_panel_decision_mi100' defined in .bss section in
get_batched_crossover.cpp:(.text.startup+0xd35): relocation truncated to fit: R_X86_64_REX_GOTPCRELX against symbol `sgeqrf_panel_decision_mi100' defined in .bss section in
get_batched_crossover.cpp:(.text.startup+0xd3c): relocation truncated to fit: R_X86_64_REX_GOTPCRELX against symbol `std::vector<std::vector<int, std::allocator<int> >, std
get_batched_crossover.cpp:(.text.startup+0xd43): relocation truncated to fit: R_X86_64_PC32 against symbol `__dso_handle' defined in .data.rel.local section in /nix/store/v
get_batched_crossover.cpp:(.text.startup+0x16ea): relocation truncated to fit: R_X86_64_REX_GOTPCRELX against symbol `dgeqrf_panel_decision_mi100' defined in .bss section i
get_batched_crossover.cpp:(.text.startup+0x1739): additional relocation overflows omitted from the output
lib/libmagma.so: PC-relative offset overflow in PLT entry for `magma_cgerc'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
lines 6661-6707/6707 (END)
```

### Notify maintainers

@tbenst 

Also @ConnorBaker @samuela may be interested.

### Metadata

```console
 - system: `"x86_64-linux"`
 - host os: `Linux 4.15.0-169-generic, Ubuntu, 18.04.6 LTS (Bionic Beaver), nobuild`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.14.0pre20230222_4a921ba`
 - channels(ben): `"home-manager, nixpkgs"`
 - nixpkgs: `/home/ben/.nix-defexpr/channels/nixpkgs`
```

--- a/pkgs/development/compilers/cudatoolkit/common.nix
+++ b/pkgs/development/compilers/cudatoolkit/common.nix
@@ -151,6 +151,10 @@ backendStdenv.mkDerivation rec {
   # Refer to comments in the overrides for cuda_nvcc for explanation
   # CUDA_TOOLKIT_ROOT_DIR is legacy,
   # Cf. https://cmake.org/cmake/help/latest/module/FindCUDA.html#input-variables
+  # NOTE: We unconditionally set -Xfatbin=-compress-all, which reduces the size of the compiled
+  #   binaries. If binaries grow over 2GB, they will fail to link. This is a problem for us, as
+  #   the default set of CUDA capabilities we build can regularly cause this to occur (for
+  #   example, with Magma).
   ''
     mkdir -p $out/nix-support
     cat <<EOF >> $out/nix-support/setup-hook
@@ -160,7 +164,7 @@ backendStdenv.mkDerivation rec {
     if [ -z "\''${CUDAHOSTCXX-}" ]; then
       export CUDAHOSTCXX=${backendStdenv.cc}/bin;
     fi
-    export NVCC_PREPEND_FLAGS+=' --compiler-bindir=${backendStdenv.cc}/bin'
+    export NVCC_PREPEND_FLAGS+=' --compiler-bindir=${backendStdenv.cc}/bin -Xfatbin=-compress-all'
     EOF
 
     # Move some libraries to the lib output so that programs that

--- a/pkgs/development/compilers/cudatoolkit/flags.nix
+++ b/pkgs/development/compilers/cudatoolkit/flags.nix
@@ -4,12 +4,8 @@
 }:
 
 # Type aliases
-# Gpu = {
-#   archName: String, # e.g., "Hopper"
-#   computeCapability: String, # e.g., "9.0"
-#   minCudaVersion: String, # e.g., "11.8"
-#   maxCudaVersion: String, # e.g., "12.0"
-# }
+# Gpu :: AttrSet
+#   - See the documentation in ./gpus.nix.
 
 let
   inherit (lib) attrsets lists strings trivial versions;
@@ -34,21 +30,39 @@ let
   # gpus :: List Gpu
   gpus = builtins.import ./gpus.nix;
 
-  # isVersionIn :: Gpu -> Bool
+  # isSupported :: Gpu -> Bool
   isSupported = gpu:
     let
       inherit (gpu) minCudaVersion maxCudaVersion;
       lowerBoundSatisfied = strings.versionAtLeast cudaVersion minCudaVersion;
-      upperBoundSatisfied = !(strings.versionOlder maxCudaVersion cudaVersion);
+      upperBoundSatisfied = (maxCudaVersion == null)
+        || !(strings.versionOlder maxCudaVersion cudaVersion);
     in
     lowerBoundSatisfied && upperBoundSatisfied;
+
+  # isDefault :: Gpu -> Bool
+  isDefault = gpu:
+    let
+      inherit (gpu) dontDefaultAfter;
+      newGpu = dontDefaultAfter == null;
+      recentGpu = newGpu || strings.versionAtLeast dontDefaultAfter cudaVersion;
+    in
+    recentGpu;
 
   # supportedGpus :: List Gpu
   # GPUs which are supported by the provided CUDA version.
   supportedGpus = builtins.filter isSupported gpus;
 
+  # defaultGpus :: List Gpu
+  # GPUs which are supported by the provided CUDA version and we want to build for by default.
+  defaultGpus = builtins.filter isDefault supportedGpus;
+
   # supportedCapabilities :: List Capability
   supportedCapabilities = lists.map (gpu: gpu.computeCapability) supportedGpus;
+
+  # defaultCapabilities :: List Capability
+  # The default capabilities to target, if not overridden by the user.
+  defaultCapabilities = lists.map (gpu: gpu.computeCapability) defaultGpus;
 
   # cudaArchNameToVersions :: AttrSet String (List String)
   # Maps the name of a GPU architecture to different versions of that architecture.
@@ -151,6 +165,6 @@ assert (formatCapabilities { cudaCapabilities = [ "7.5" "8.6" ]; }) == {
   # dropDot :: String -> String
   inherit dropDot;
 } // formatCapabilities {
-  cudaCapabilities = config.cudaCapabilities or supportedCapabilities;
+  cudaCapabilities = config.cudaCapabilities or defaultCapabilities;
   enableForwardCompat = config.cudaForwardCompat or true;
 }

--- a/pkgs/development/compilers/cudatoolkit/gpus.nix
+++ b/pkgs/development/compilers/cudatoolkit/gpus.nix
@@ -1,110 +1,148 @@
 [
+  # Type alias
+  # Gpu = {
+  #   archName: String
+  #     - The name of the microarchitecture.
+  #   computeCapability: String
+  #     - The compute capability of the GPU.
+  #   minCudaVersion: String
+  #     - The minimum (inclusive) CUDA version that supports this GPU.
+  #   dontDefaultAfter: null | String
+  #     - The CUDA version after which to exclude this GPU from the list of default capabilities
+  #       we build. null means we always include this GPU in the default capabilities if it is
+  #       supported.
+  #   maxCudaVersion: null | String
+  #     - The maximum (exclusive) CUDA version that supports this GPU. null means there is no
+  #       maximum.
+  # }
   {
     archName = "Kepler";
     computeCapability = "3.0";
     minCudaVersion = "10.0";
+    dontDefaultAfter = "10.2";
     maxCudaVersion = "10.2";
   }
   {
     archName = "Kepler";
     computeCapability = "3.2";
     minCudaVersion = "10.0";
+    dontDefaultAfter = "10.2";
     maxCudaVersion = "10.2";
   }
   {
     archName = "Kepler";
     computeCapability = "3.5";
     minCudaVersion = "10.0";
+    dontDefaultAfter = "11.0";
     maxCudaVersion = "11.8";
   }
   {
     archName = "Kepler";
     computeCapability = "3.7";
     minCudaVersion = "10.0";
+    dontDefaultAfter = "11.0";
     maxCudaVersion = "11.8";
   }
   {
     archName = "Maxwell";
     computeCapability = "5.0";
     minCudaVersion = "10.0";
-    maxCudaVersion = "12.0";
+    dontDefaultAfter = "11.0";
+    maxCudaVersion = null;
   }
   {
     archName = "Maxwell";
     computeCapability = "5.2";
     minCudaVersion = "10.0";
-    maxCudaVersion = "12.0";
+    dontDefaultAfter = "11.0";
+    maxCudaVersion = null;
   }
   {
     archName = "Maxwell";
     computeCapability = "5.3";
     minCudaVersion = "10.0";
-    maxCudaVersion = "12.0";
+    dontDefaultAfter = "11.0";
+    maxCudaVersion = null;
   }
   {
     archName = "Pascal";
     computeCapability = "6.0";
     minCudaVersion = "10.0";
-    maxCudaVersion = "12.0";
+    dontDefaultAfter = null;
+    maxCudaVersion = null;
   }
   {
     archName = "Pascal";
     computeCapability = "6.1";
     minCudaVersion = "10.0";
-    maxCudaVersion = "12.0";
+    dontDefaultAfter = null;
+    maxCudaVersion = null;
   }
   {
     archName = "Pascal";
     computeCapability = "6.2";
     minCudaVersion = "10.0";
-    maxCudaVersion = "12.0";
+    dontDefaultAfter = null;
+    maxCudaVersion = null;
   }
   {
     archName = "Volta";
     computeCapability = "7.0";
     minCudaVersion = "10.0";
-    maxCudaVersion = "12.0";
+    dontDefaultAfter = null;
+    maxCudaVersion = null;
   }
   {
     archName = "Volta";
     computeCapability = "7.2";
     minCudaVersion = "10.0";
-    maxCudaVersion = "12.0";
+    dontDefaultAfter = null;
+    maxCudaVersion = null;
   }
   {
     archName = "Turing";
     computeCapability = "7.5";
     minCudaVersion = "10.0";
-    maxCudaVersion = "12.0";
+    dontDefaultAfter = null;
+    maxCudaVersion = null;
   }
   {
     archName = "Ampere";
     computeCapability = "8.0";
     minCudaVersion = "11.2";
-    maxCudaVersion = "12.0";
+    dontDefaultAfter = null;
+    maxCudaVersion = null;
   }
   {
     archName = "Ampere";
     computeCapability = "8.6";
     minCudaVersion = "11.2";
-    maxCudaVersion = "12.0";
+    dontDefaultAfter = null;
+    maxCudaVersion = null;
   }
   {
     archName = "Ampere";
     computeCapability = "8.7";
     minCudaVersion = "11.5";
-    maxCudaVersion = "12.0";
+    # NOTE: This is purposefully before 11.5 to ensure it is never a capability we target by
+    #   default. 8.7 is the Jetson Orin series of devices which are a very specific platform.
+    #   We keep this entry here in case we ever want to target it explicitly, but we don't
+    #   want to target it by default.
+    dontDefaultAfter = "11.4";
+    maxCudaVersion = null;
   }
   {
     archName = "Ada";
     computeCapability = "8.9";
     minCudaVersion = "11.8";
-    maxCudaVersion = "12.0";
+    dontDefaultAfter = null;
+    maxCudaVersion = null;
   }
   {
     archName = "Hopper";
     computeCapability = "9.0";
     minCudaVersion = "11.8";
-    maxCudaVersion = "12.0";
+    dontDefaultAfter = null;
+    maxCudaVersion = null;
   }
 ]

--- a/pkgs/development/compilers/cudatoolkit/redist/overrides.nix
+++ b/pkgs/development/compilers/cudatoolkit/redist/overrides.nix
@@ -41,6 +41,10 @@ in
       # uses the last --compiler-bindir it gets on the command line.
       # FIXME: this results in "incompatible redefinition" warnings.
       # https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#compiler-bindir-directory-ccbin
+      # NOTE: We unconditionally set -Xfatbin=-compress-all, which reduces the size of the
+      #   compiled binaries. If binaries grow over 2GB, they will fail to link. This is a problem
+      #   for us, as the default set of CUDA capabilities we build can regularly cause this to
+      #   occur (for example, with Magma).
       postInstall = (oldAttrs.postInstall or "") + ''
         mkdir -p $out/nix-support
         cat <<EOF >> $out/nix-support/setup-hook
@@ -49,7 +53,7 @@ in
         if [ -z "\''${CUDAHOSTCXX-}" ]; then
           export CUDAHOSTCXX=${cc}/bin;
         fi
-        export NVCC_PREPEND_FLAGS+=' --compiler-bindir=${cc}/bin'
+        export NVCC_PREPEND_FLAGS+=' --compiler-bindir=${cc}/bin -Xfatbin=-compress-all'
         EOF
       '';
     });

